Sources/Confidence/EventStorage.swift

```diff
@@ -0,0 +1,127 @@
+                return
+            }
+            currentFileHandle.seekToEndOfFile()
+            try currentFileHandle.write(contentsOf: delimiter)
+            try currentFileHandle.write(contentsOf: serialied)
+        }
+    }
+
+
+    func batchReadyIds() throws -> [String] {
+        try storageQueue.sync {
+            let fileUrls = try FileManager.default.contentsOfDirectory(at: folderURL, includingPropertiesForKeys: nil)
+            return fileUrls.filter({ url in url.pathExtension == READYTOSENDEXTENSION }).map({ url in url.lastPathComponent })
+        }
+    }
+
+    func eventsFrom(id: String) throws -> [Event] {
+        try storageQueue.sync {
+            let decoder = JSONDecoder()
+            let fileUrl = folderURL.appendingPathComponent(id)
+            let data = try Data(contentsOf: fileUrl)
+            let dataString = String(data: data, encoding: .utf8)
+            return try dataString?.components(separatedBy: "\n")
+                .filter({ events in !events.isEmpty })
+                .map({ eventString in try decoder.decode(Event.self, from: eventString.data(using: .utf8)!) }) ?? []
+        }
+    }
+
+    func remove(id: String) throws {
+        try storageQueue.sync {
+                let fileUrl = folderURL.appendingPathComponent(id)
+                try FileManager.default.removeItem(at: fileUrl)
+        }
+    }
+
+    private func getLastWritingFile() throws -> URL? {
+        let files = try FileManager.default.contentsOfDirectory(at: folderURL, includingPropertiesForKeys: nil)
+        for fileUrl in files {
+            if fileUrl.pathExtension != READYTOSENDEXTENSION {
+                return fileUrl
+            }
+        }
+        return nil
+    }
+
+    private func resetCurrentFile() throws {
+        if let currentFile = try getLastWritingFile() {
+            self.currentFileUrl = currentFile
+            self.currentFileHandle = try FileHandle(forWritingTo: currentFile)
+        } else {
+            let fileUrl = folderURL.appendingPathComponent(String(Date().timeIntervalSince1970))
+            FileManager.default.createFile(atPath: fileUrl.path, contents: nil)
+            self.currentFileUrl = fileUrl
+            self.currentFileHandle = try FileHandle(forWritingTo: fileUrl)
+        }
+    }
+
+    internal static func getFolderURL() throws -> URL {
+        guard
+            let applicationSupportUrl = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)
+                .last
+        else {
+            throw ConfidenceError.cacheError(message: "Could not get URL for application directory")
+        }
+
+        guard let bundleIdentifier = Bundle.main.bundleIdentifier else {
+            throw ConfidenceError.cacheError(message: "Unable to get bundle identifier")
+        }
+
+        return applicationSupportUrl.backport.appending(
+            components: "com.confidence.cache", "\(bundleIdentifier)", "events")
+    }
```

Comment:
Not sure what extra value the `"\(bundleIdentifier)",` brings?

Reply:
looked it up [here](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW11). according to apple we need the unique `bundleIdentifier` when working with application support directory.

Reply:
What a wonderful API. 😞